lib/vector-buffers/src/disk_v2/reader.rs

```diff
@@ -305,7 +329,17 @@ where
         // - `try_next_record` does all the archive checks, checksum validation, etc
         let archived_record = unsafe { archived_root::<Record<'_>>(&self.aligned_buf) };
 
-        T::decode(archived_record.payload()).context(DecodeSnafu)
+        // Try and convert the raw record metadata into the true metadata type used by `T`, and then
+        // also verify that `T` is able to decode records with the metadata used for this record in particular.
+        let metadata =
+            T::Metadata::from_u32(archived_record.metadata()).ok_or(ReaderError::Incompatible)?;
```

Comment:
Would we ever want separate errors between this step and the `can_decode` step?

Reply:
We might! I did have that thought, or at least the thought that there could bean additional `reason` field describing the "this metadata isn't even valid" vs "we don't support this record".  Maybe I'll just go ahead and implement that.

Reply:
Addressed this in 006a597.